dask/array/core.py

```diff
@@ -989,6 +989,13 @@ def finalize(results):
     return unpack_singleton(results)
 
 
+def finalize(results):
+    res = _finalize(results)
+    if type(res) == np.ndarray and not res.flags.writeable:
+        res = np.array(res, copy=True)
```

Comment:
Would suggest using the `.copy()` method so this works with `sparse` and `cupy` arrays as well.

```suggestion
        res = res.copy()
```

Reply:
It's up to you what you think we should do, so I tried to restrict myself to only copy the case introduces by zeros/ones/full, so I don't see any reason to copy those. I've pushed [another branch](https://travis-ci.org/github/Carreau/dask/builds/719613905) to see if `array.copy()` pass the tests.